### DISCUSSION TITLE
Health checks for chiseled containers

### DISCUSSION THREAD

**Discussion by MikaelElkiaer** [dotnet/dotnet-docker#4296] [Q&A] [answered]
Hiya, great work on the chiseled containers, looking forward for them to be GA.

I am using dotnet Docker images for ASP.NET apps deployed in both Kubernetes and simpler Docker-based set-ups.
For Kubernetes, these chiseled images are awesome, so far without any issues.

However, for Docker (docker-compose) based deployments, I would like to utilize the built-in health checks.
Does anyone have any thoughts on how to do built-in health checks, for an ASP.NET application running in a chiseled image?

**Comment by mthalman** [comment, marked as answer]
How do you currently do health checks and what is it about chiseled containers that makes that approach problematic?

**Reply by MikaelElkiaer** [reply]
For ASP.NET, I at least have the [basic health probe](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-7.0#basic-health-probe) configured:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddHealthChecks();

var app = builder.Build();

app.MapHealthChecks("/healthz");

app.Run();
```
 
Then I will include this simple [Docker health check](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-7.0#docker-healthcheck): 

```csharp
HEALTHCHECK CMD curl --fail http://localhost:5000/healthz || exit
```

Would I then have to extend the chiseled image to include curl, wget, or similar?
Or is there another way?

**Reply by mthalman** [reply]
Thanks for explaining your scenario. This is a general problem of chiseled/distroless containers. There's a good discussion on this topic at https://github.com/GoogleContainerTools/distroless/issues/183 with various solutions. Personally, I like the idea of having a small custom healthcheck .NET app that would get called to run the HTTP request.

**Reply by MikaelElkiaer** [reply]
Thanks - good discussion in the issue that you linked.
I ended up doing something similar to this: https://github.com/GoogleContainerTools/distroless/issues/183#issuecomment-1273120299

A simple `./healthchecker/Program.cs`:

```csharp
var httpClient = new HttpClient();
httpClient.DefaultRequestHeaders.ConnectionClose = true;

if (args.Length > 1 && Uri.TryCreate(args[1], UriKind.Absolute, out var uri))
  return await httpClient.GetAsync(uri).ContinueWith(r => r.Result.IsSuccessStatusCode) ? 0 : 1;
else
  throw new ArgumentNullException("A valid URI must be given as first argument");
```

And a publish of that:

`dotnet publish --configuration Release --runtime linux-x64 --no-self-contained healthchecker -o healthchecker/out`

And finally the Docker health check:

```Dockerfile
# ...

COPY ./healthchecker/out /healthchecker

HEALTHCHECK CMD [ "dotnet", "/healthchecker/healthchecker.dll", "--", "http://localhost:8080/healthz" ]
```

**Reply by mthalman** [reply]
I created this issue with a proposal of defining a shared tool that could provide this functionality: https://github.com/dotnet/dotnet-docker/issues/4300

**Reply by lkroliko** [reply]
Hey. I had the same problem. I made nuget package for easy use. 
[MrRabbit.HealthChecks.Container.Client](https://www.nuget.org/packages/MrRabbit.HealthChecks.Container.Client/)

**Reply by zharchimage** [reply]
Hi I´m struggling also with chiseled asp.net images. I also did a small dotnet app that I compile with every dockerfile and include it as a sidecar with every service that I need to healthcheck. Although this approach is working, I´m failing to log information from the healthcheck dll as it runs in its own process and can´t print useful logs to the main docker logs. Console.WriteLine nor Debug.Writeline work. Any idea who could we redirect the healthchecker.dll output to the main log in a docker compose healthcheck? Thanks!

**Reply by lkroliko** [reply]
> Hi I´m struggling also with chiseled asp.net images. I also did a small dotnet app that I compile with every dockerfile and include it as a sidecar with every service that I need to healthcheck. Although this approach is working, I´m failing to log information from the healthcheck dll as it runs in its own process and can´t print useful logs to the main docker logs. Console.WriteLine nor Debug.Writeline work. Any idea who could we redirect the healthchecker.dll output to the main log in a docker compose healthcheck? Thanks!

Hello. The container console is connected to your application running in the container, so an easy way is to add logs in the application rather than in the healthchecker using an API call. Docker healthchecks runs in shell and their output can be found in the 'output' property when you execute 'docker inspect [container id]'.
```json
{
 "State": {
       
            "Health": {
                "Status": "healthy",
                "FailingStreak": 0,
                "Log": [
                    {
                        "Start": "2025-09-29T11:31:10.747255446+02:00",
                        "End": "2025-09-29T11:31:10.966375299+02:00",
                        "ExitCode": 0,
                        "Output": "[INF] - Success response status code 'OK.\n"
                    }
                ]
            }
        }
}
```

**Reply by zharchimage** [reply]
Yes, I´m already doing that, however I´m trying to plug in some healthcheck and they are failing for some reason, for example, the URL I´m passing is not the one the application is running on, that is why it is useful for me to check the log status for the healthchecks. I´m guessing maybe there is a command to redirect the stdout from the healthchecks to the main process where the docker application is running? I´m trying something like this

`    test: ["CMD", "dotnet" , "/app/healthcheck/HealthCheckService.dll" , ">" ,"/dev/stdout"]`

**Comment by dustinbrooks** [comment]
I feel like I'm missing something trying to get the small dotnet healthcheck app working. Since the chiseled container doesn't have a shell, how does it execute the HEALTHCHECK CMD? 
I'm utilizing the 8.0-noble-chiseled image. Here is my healthcheck, which I can run this command successfully if I use docker desktop's debug mode to spin up a shell.

`HEALTHCHECK CMD dotnet /app/healthcheck.dll http://localhost:8080/Health/Live || exit 1`

Here is my docker inspect on the health check:

`{
  "Status": "unhealthy",
  "FailingStreak": 12,
  "Log": [
    {
      "Start": "2024-12-05T14:49:29.714229238Z",
      "End": "2024-12-05T14:49:29.746231576Z",
      "ExitCode": -1,
      "Output": "OCI runtime exec failed: exec failed: unable to start container process: exec: \"/bin/sh\": stat /bin/sh: no such file or directory: unknown"
    },`

To me that reads like it can't run the dotnet command because it can't find the shell to do so.

**Reply by mthalman** [reply]
The difference lies in the syntax you're using. In the [solution above](https://github.com/dotnet/dotnet-docker/discussions/4296#discussioncomment-4602902), the JSON array syntax was used for the `HEALTHCHECK` instruction. This does not require a shell to execute it; rather, Docker executes the command directly. In your command, it uses the shell form syntax which does require a shell in the container to execute. So you would want to use the JSON array syntax instead. You're limited what you can do there, so you can't add the `|| exit 1` portion as that is shell syntax.

**Reply by lbussell** [reply]
Yes, here is the Dockerfile documentation on the "shell" and "exec" forms of instructions: https://docs.docker.com/reference/dockerfile/#shell-and-exec-form

**Reply by MikaelElkiaer** [reply]
Some time ago I made this: https://github.com/MikaelElkiaer/dotnet-healthchecker

It might need some love and care now, but perhaps it is still usable.

**Reply by dustinbrooks** [reply]
@mthalman @lbussell Thank you. Much appreciated. Changing my format to exec form was exactly what I needed to learn about and use.
